### PR TITLE
Reuse prepared statements for repeat statements in unnamed mode

### PR DESCRIPTION
**Pull request by hkrutzer**

### CODE REVIEW

File: test/myxql_test.exs [hunk 1]
```diff
@@ -189,7 +189,8 @@ defmodule MyXQLTest do
       {:ok, pid} = MyXQL.start_link(@opts ++ [prepare: :unnamed])
       {:ok, query} = MyXQL.prepare(pid, "1", "SELECT 1")
       {:ok, query2, _} = MyXQL.execute(pid, query, [])
-      assert query == query2
+      assert query.ref != query2.ref
+      assert query.statement_id != query2.statement_id
```
Comment (hkrutzer) [inline test/myxql_test.exs:193]: Not 100% sure about this one

**Comment by hkrutzer** [comment]
@wojtekmach do you have time to take a look at this?

**Comment by wojtekmach** [comment]
Yeah, sorry for radio silence, I'll try to get to it soon.

**Comment by hkrutzer** [comment]
@wojtekmach @josevalim ping :)

File: lib/myxql/connection.ex [hunk 2]
```diff
@@ -107,9 +111,17 @@ defmodule MyXQL.Connection do
...
-      end
+      state =
+        case result do
+          {:ok, err_packet()} ->
+            {:ok, state} = close(query, state)
+            state
+
+          _ ->
+            state
+        end
```
Comment (josevalim) [inline lib/myxql/connection.ex:122]:
```suggestion
      {:ok, state} =
        case result do
          {:ok, err_packet()} -> close(query, state)
          _ -> {:ok, state}
        end
```

File: test/myxql/sync_test.exs [hunk 3]
```diff
@@ -9,7 +9,33 @@ defmodule MyXQL.SyncTest do
...
+  end
+
+  test "produce the expected amount of prepared statements with prepare: :named" do
+    {:ok, conn} = MyXQL.start_link(@opts)
     assert prepared_stmt_count() == 0
+
+    MyXQL.query!(conn, "SELECT 42", [], cache_statement: "42")
+    assert prepared_stmt_count() == 1
+
+    MyXQL.query!(conn, "SELECT 1337", [], cache_statement: "69")
```
Comment (josevalim) [inline test/myxql/sync_test.exs:28]:
```suggestion
    MyXQL.query!(conn, "SELECT 1337", [], cache_statement: "1337")
```

File: test/myxql_test.exs (hunk 1 above)
Comment (josevalim) [reply test/myxql_test.exs:194]: We should be executing the prepared query without repreparing it.

**Comment by greg-rychlewski** [comment]
Is the idea here that you if you `execute` the same unnamed query more than once in a row, you don't re-prepare? But if you are using `query` or `prepare` or `prepare_execute` then you always re-prepare so you're not comparing statements each time you issue a new unnamed query?

Just asking because I had some ideas reading through this PR but want to make sure I understand the goal.

**Comment by josevalim** [comment]
Yes, that's the idea. note it applies only to unnamed mode (when we prepare the queries and immediately discard them).

**Comment by greg-rychlewski** [comment]
I think this does what you want: https://github.com/greg-rychlewski/myxql/commit/fd99c3009c3d82b90a5535b2e9e8302299b7cfee

For unnamed:
- don't immediately close the query
- close the last query on new prepare or executing a different query

For named:
- keep the same behaviour immediately closing queries with `""` name (i.e. if they call `query` without specifying `cache_statement`). the alternative is requiring every named query to check if the last query has name `""` and potentially issuing 2 closing messages (the `""` query and a stale cache entry)

Sorry I put this in my own branch because I thought it would be clearer than issuing fragmented comments. But if these ideas are good they can be applied here.

**Comment by greg-rychlewski** [comment]
@josevalim I didn't notice the age of this PR the first time I replied. Would it be acceptable to submit my version here: greg-rychlewski@fd99c30? Or if this is expected to pick back up I can leave it alone, no problem.

**Comment by josevalim** [comment]
@greg-rychlewski you can also send your commit on top of this one. So everyone gets recognized. Would that work?

**Comment by greg-rychlewski** [comment]
Oh yeah, for sure. Sorry I just never did anything like that before. I just push directly to the fork being used by the PR?

**Comment by josevalim** [comment]
Here is what I would do:

1. Get @hkrutzer's branch on your machine
2. Apply your commit on top of his
3. Push this new branch from your fork
4. Send a PR

**Comment by wojtekmach** [comment]
Closing in favour of #146

**Pull request closed without merging**